server/core/storelimit/sliding_window.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2022 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storelimit
+
+import (
+	"github.com/tikv/pd/pkg/utils/syncutil"
+)
+
+const (
+	// minSnapSize is the min value to check the windows has enough size.
+	minSnapSize = 10
+)
+
+var _ StoreLimit = &SlidingWindows{}
+
+// SlidingWindows is a multi sliding windows
+type SlidingWindows struct {
+	mu      syncutil.RWMutex
+	windows []*window
+}
+
+// NewSlidingWindows is the construct of SlidingWindows.
+func NewSlidingWindows(capacity int64) *SlidingWindows {
+	if capacity < 0 {
+		capacity = minSnapSize
+	}
+	windows := make([]*window, priorityLevelLen)
+	for i := 0; i < int(priorityLevelLen); i++ {
+		windows[i] = newWindow(capacity >> i)
+	}
+
+	return &SlidingWindows{
+		windows: windows,
+	}
+}
+
+// Reset resets the capacity of the sliding windows.
+func (s *SlidingWindows) Reset(cap float64, _ Type) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if cap < 0 {
+		cap = minSnapSize
+	}
+	for i, v := range s.windows {
+		v.reset(int64(cap) >> i)
+	}
+}
+
+// GetUsed returns the used size in the sliding windows.
+func (s *SlidingWindows) GetUsed() int64 {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	used := int64(0)
+	for _, v := range s.windows {
+		used += v.getUsed()
+	}
+	return used
+}
+
+// Available returns whether the token can be taken.
+// The order of checking windows is from low to high.
+// It checks the given window finally if the lower window has no free size.
+func (s *SlidingWindows) Available(_ int64, _ Type, level PriorityLevel) bool {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	for i := 0; i < int(level); i++ {
+		if s.windows[i].available() {
+			return true
+		}
+	}
+	return false
+}
+
+// Take tries to take the token.
+// It will consume the given window finally if the lower window has no free size.
+func (s *SlidingWindows) Take(token int64, _ Type, level PriorityLevel) bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	for i := 0; i < int(level); i++ {
```

Comment:
It seems that it always returns false if level is `Low`?

Reply:
yes, I will fix it with `<=`